refactor(zone.js): monkey patch async APIs with unified `patchMethod` API.

JiaLiPassion opened the pull request:
Refactor zone.js to use a unified way to do the `monkey patch`.

In the current version, we are doing the monkey patching differently. 
For example, 

1. `setTimeout/XHR/..`, we are using something like
```
const original = window.setTimeout;
window.setTimeout = function(callback) {
  original.call(window, Zone.current.wrap(callback),...);
}
```

2. For `FileReader` and some other ones, we are using `patchClass`.

```
class PatchedFileReader {
  constructor() { 
    const f = new OriginalFileReader();
    for (prop in f) {
      // if prop is function, wrap it,
     wrapFunctionWithCurrentZone(...);
    }
  }

  //
`}
```

3. For `Promise`, and some other complex ones, we are monkey patching them with own logic.
4. And there are some other cases such as `EventTarget/toString/...`, they all have some different monkey patching logic.

It is working, but very difficult to maintain.

So this PR is doing a refactor to make all monkey patching use the unified `patchMethod()` to do the monkey patching.

The new monkey patching looks like this.

```
const original = window.setTimeout;
Object.defineProperty(window, 'setTimeout', {
  get: function() {
    return patchedSetTimeout;  
  };
  set: function() {...}
})
```

So we are using an `Object.defineProperty` to do the monkey patch, so we can do more extensions on the monkey patching if we want easier.

And the first extension is `enable/disable` zone.js on the fly.

```
const original = window.setTimeout;
Object.defineProperty(window, 'setTimeout', {
  get: function() {
    return Zone.isPatchEnabled() ? patchedSetTimeout : nativeSetTimeout;  
  };
  set: function() {...}
})
```

Now we can 

```
Zone.current.fork({
  name: 'testZone'
}).run(() => {
  setTimeout(() => {
    console.log('timeout callback', Zone.current.name); // this will output testZone since zone.js is enabled by default
  }); 
  Zone.disablePatch();
  setTimeout(() => {
    console.log('timeout callback', Zone.current.name); // this will output <root> since zone.js is diabled.
  }); });
  Zone.enablePatch();
  setTimeout(() => {
    console.log('timeout callback', Zone.current.name); // this will output testZone again since zone.js is enabled.
  }); });

```

So we can do something fun with this new APIs.

JiaLiPassion committed:
refactor(zone.js): patch callback with api.patchMethod()

Use `api.patchMethod()` to patch callbacks, so we can use the unified
method to do monkey patch. Instead of using some unique way to do it.

JiaLiPassion committed:
refactor(zone.js): remove FileReader patch.

Don't need to patch FileReader any longer since it is an EventTarget.

JiaLiPassion committed:
refactor(zone.js): remove websocket patch.

Don't need to do websocket monkey patch since it is an EventTarget.

JiaLiPassion committed:
refactor(zone.js): remove websocket patch.

Remove websocket patch since Websocket is an EventTarget, and we
don't need to support old browsers any longer.

JiaLiPassion committed:
refactor(zone.js): patch WebSocket/FileReader onProperties.

Since we remove the WebSocket/FileReader patch, we need to patch
some of the `on` properties such as `onmessage`.

JiaLiPassion committed:
refactor(zone.js): use patchMethod patch userMedia.

Refactor to use the unified `patchMethod` to patch `userMedia`.

JiaLiPassion committed:
refactor(zone.js): update FileReader test case.

Update FileReader test cases to make it more readable.

JiaLiPassion committed:
refactor(zone.js): use patchMethod to patch EventTarget

Refactor use unified `patchMethod` to patch `EventTarget`.

JiaLiPassion committed:
refactor(zone.js): use patchMethod for patchClass

Refactor, use the unified `patchMethod` to patchClass.

JiaLiPassion committed:
feat(zone.js): add patch flag and enable/disable patch method.

Add `enablePatch()`/`disablePatch()` function.

JiaLiPassion committed:
feat(zone.js): add new api to zone_externs.js

Add new `enablePatch()/disablePatch()` to `zone_externs.js` to ensure
google closure work.

JiaLiPassion committed:
refactor(zone.js): use Object.defineProperty to patchMethod

Use `Object.defineProperty` instead set the value directly in
`patchMethod()` monkey patch.

JiaLiPassion committed:
refactor(zone.js): use patchFunctionProperty to patch Error.

Refactor: use `patchFunctionProperty` (internally is patchMethod()` to
patch `global.Error`.

JiaLiPassion committed:
refactor(zone.js): use patchFunctionProperty to patch fetch.

Refactor: use `patchFunctionProperty` to patch `globa.fetch`.

JiaLiPassion committed:
refactor(zone.js): use patchFunctionProperty to patch Promise.

Refactor, Use `patchFunctionProperty()` (internally patchMethod()` to
patch Promise.

JiaLiPassion committed:
refactor(zone.js): use patchMethod to patch to-string.

Refactor: use patchMethod() to patch `Object.prototype.toString()`

JiaLiPassion committed:
fix(zone.js): fallback should also check isPatchEnabled.

Should use the original native delegate when patch is disabled.

JiaLiPassion committed:
refactor(zone.js): node patch use patchMethod/patchFunctionProperty.

Refactor: use the unified `patchMethod` to patch nodejs APIs.

JiaLiPassion committed:
test(zone.js): add scoped zone test cases

Add test cases for scoped zone.

Commit:
test(zone.js): update fetch spec

Update fetch spec since the scoped zone implementation.

JiaLiPassion committed:
test(zone.js): update scopepd zone test cases

Test: Add more test cases for async API test of scoped zone.

elgreco247 commented on packages/zone.js/lib/zone.ts:
```diff
@@ -344,6 +359,7 @@ interface _ZonePrivate {
       (target: any, name: string,
        patchFn: (delegate: Function, delegateName: string, name: string) =>
            (self: any, args: any[]) => any) => Function | null;
+  patchFunctionProprety(target: any, name: string, patchFn: Function): boolean;
```
typo: should be 'patchFunctionProperty'

JiaLiPassion replied:
@elgreco247 , yeah, you are right, thank you.

JiaLiPassion committed:
fix(zone.js): fix typo

Fix typo error for patchFunctionProperty method name.

JiaLiPassion committed:
fix(zone.js): update golden files

Update golden files since zone.js size decreased.

lmatejic9 commented:
Hi, would just like to check if this has some issue linked to it to follow ?
Would be interested in seeing this sometime in the future.

JiaLiPassion committed:
fix(zone.js): customElements patch should not patch prototype

In the previous version, `customElements.define` is patched on the
window.customElements itself instead of CustomElementsRegistry, so
we should not patch on the CustomElementsRegistry.
